### ISSUE TITLE
DDP only syncs parameters used in most recent pass when `find_unused_parameters` is True.

### ISSUE THREAD

**Issue by jbloxham** [pytorch/pytorch#69031]
## 🐛 Bug

I'm working with a training loop with gradient accumulation, where several forwards/backwards passes run for each call to the optimizer step. To reduce communication overhead, DDP syncs are disabled for all but the final gradient accumulation step before the optimizer step. I'm also using an optimized model which randomly disables various parts of itself on each pass. Critically, the parts of the model that are disabled can vary between gradient accumulation steps. The problem I'm observing is that the DDP sync ran on that final grad accum step appears to only be syncing the parameters used on that step. Parameters unused on the final step but used on previous steps are never synced.

## To Reproduce

```python
import contextlib
import os

import torch
import torch.distributed
import torch.multiprocessing
import torch.nn as nn

GRADIENT_ACCUMULATION_STEPS = 2


class SimpleConditionalModel(nn.Module):
    # uses nn1 layer on the first pass; nn2 layer on the second pass

    def __init__(self):
        super().__init__()

        self.nn1 = nn.Linear(1, 1)
        self.nn2 = nn.Linear(1, 1)

        self.loss = nn.MSELoss()

        self.state = 0

    def forward(self, input):
        if self.state == 0:
            self.state = 1
            return self.nn1(input)
        else:
            self.state = 0
            return self.nn2(input)


def test_model(rank):
    os.environ["RANK"] = str(rank)
    os.environ["MASTER_ADDR"] = "127.0.0.1"
    os.environ["MASTER_PORT"] = "29000"

    torch.distributed.init_process_group(backend='gloo', world_size=2)
    model = SimpleConditionalModel()
    ddp = torch.nn.parallel.DistributedDataParallel(model, find_unused_parameters=True)

    for microbatch_idx in range(GRADIENT_ACCUMULATION_STEPS):
        context = contextlib.nullcontext
        if microbatch_idx < GRADIENT_ACCUMULATION_STEPS - 1:
            context = ddp.no_sync

        with context():
            input = torch.rand((1, ))
            output = ddp.forward(input)
            target = torch.rand((1, ))

            loss = model.loss(output, target)
            loss.backward()

    print(f'Rank {rank} gradients: {[p.grad for p in model.parameters()]}')


if __name__ == '__main__':
    torch.multiprocessing.spawn(test_model, nprocs=2)
```

Output:
```
Rank 0 gradients: [tensor([[-1.1942]]), tensor([-1.4291]), tensor([[-0.4950]]), tensor([-3.3037])]
Rank 1 gradients: [tensor([[-0.3461]]), tensor([-0.4722]), tensor([[-0.4950]]), tensor([-3.3037])]
```
The first two gradients, corresponding to the `nn1` layer, are unsynced.

## Expected behavior

DDP should be able to track used/unused parameters across multiple forwards/backwards passes.

## Environment
 - PyTorch Version: 1.9.0
 - OS: Linux
 - How you installed PyTorch: `pip`
 - Python version: 3.8.0

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang

**Comment by wanchaol** [comment]
@jbloxham I think this is because find_unused_parameters option will take the input and feed into DDP to detect the unused parameters, so it will at least need to go through the whole model (not a part of it) to detect all unused parameters

**Comment by rohan-varma** [comment]
This behavior is probably expected, though should be fixed considering we want to ensure the models are synchronized across ranks when gradient sync is enabled.

I think this happens because DDP will ensure that if a parameter is unused globally (across all ranks), it will not touch the parameter gradient, as seen here: https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L1345, https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/reducer.cpp#L1448

As a result, the .grad field is not populated with the allreduced result if the parameter was unused during the entire no_sync lifetime.

The original motivation for this was to make gradient zero when it should be None. Although I think we can fix this by checking if the grad is not None, @mrshenli / @zhaojuanmao any thoughts?

**Comment by rohan-varma** [comment]
Fixing this can also be a good ramp up/onboarding task.

**Comment by nkyle04** [comment]
I enconter this bug too. It was working well when using pytorch1.8, but fails at pytorch >1.9
And it looks like a bug at https://github.com/pytorch/pytorch/blob/845a82b635c3e3deb54f2c8baf4f997731cea3ed/torch/csrc/distributed/c10d/reducer.cpp#L565
which i this check expect_autograd_hooks_ flag should be moved to after setting local_used_map.

**Comment by TNTWEN** [comment]
Hi！
@rohan-varma  
Is there any alternative solution to this issues? It seems that torch (Preview Nightly)still can't track used/unused parameters.
